pkg/pub_worker/lib/src/http.dart

```diff
@@ -0,0 +1,47 @@
+import 'dart:async' show FutureOr;
+import 'package:http/http.dart' as http;
+
+/// Returns an [http.Client] which sends a `Bearer` token as `Authorization`
+/// header for each request.
+extension WithAuthorization on http.Client {
+  http.Client withAuthorization(
+    FutureOr<String?> Function() tokenProvider, {
+    bool closeParent = false,
+  }) =>
+      _AuthenticatedClient(
+        tokenProvider,
+        this,
+        closeParent,
+      );
+}
+
+/// An [http.Client] which sends a `Bearer` token as `Authorization` header for
+/// each request.
+class _AuthenticatedClient extends http.BaseClient {
+  final FutureOr<String?> Function() _tokenProvider;
+  final http.Client _client;
+  final bool _closeInnerClient;
+
+  _AuthenticatedClient(
+    this._tokenProvider,
+    this._client,
+    this._closeInnerClient,
+  );
+
+  @override
+  Future<http.StreamedResponse> send(http.BaseRequest request) async {
+    final token = await _tokenProvider();
+    if (token != null) {
+      request.headers['Authorization'] = 'Bearer $token';
```

Comment:
Can you elaborate how is `http_client` better? I agree that this annoying, but we could also just attach to the requests manually, making a client that injects it, is a bit of a hack...

And also a bit risky, because what if the server redirects, this just blindly attaches the token to everything :see_no_evil: 

In this case, I think it's fine, but in general, maybe it's not the best solution for injecting authentication.

Reply:
Yes, breaking this into a separate package would be nice, if you have a good package name -- I'm happy to accept this as a contribution to dart-neats.

Adding it to `package:http` requires we are sure this is the design. Maybe it is, I'm just not entirely sure about wrapping clients and `closeParent` stuff..